src/frontend/packages/core/src/features/home/home/home-page.component.ts

```diff
@@ -48,22 +94,209 @@ export class HomePageComponent {
+    }
+    setTimeout(() => this.checkCardsInView(), 1);
+  }
+
+  // This is called after a card has loaded - we call the scroll handler again
+  // to check if there are more cards that are visible and thus can be loaded
+  cardLoaded() {
+    this.isLoadingACard = false;
+    this.processCardsToLoad();
+    this.checkCardsInView();
+  }
+
+  @HostListener('window:resize')
+  onResize() {
+    // If we resize the window and make it larger then new cards may come into view
+    this.checkCardsInView();
+  }
+
+  // Check the cards in view
+  checkCardsInView() {
+    this.checkLayout.next(true);
+  }
+
+  // The layout was changed
+  public onChangeLayout(layout: HomePageCardLayout) {
+    this.layoutID = layout.id;
+
+    // If the layout is automatic, then adjust based on number of things to show
+    const lay$ = layout.id === 0 ? this.automaticLayout() : of(layout);
+    lay$.pipe(first()).subscribe(lo => {
+      this.layout.next(lo);
+
+      // Update the grid columns based on the layout
+      this.columns = lo.x;
+
+      // Persist the state
+      this.store.dispatch(new SetHomeCardLayoutAction(this.layoutID));
+
+      // Ensure we check again if any cards are now visible
+      // Schedule the check so it happens afer the cards have been laid out
+      setTimeout(() => this.checkCardsInView(), 1);
+    });
+  }
+
+  // Order the endpoint cards - we always show all endpoints, order is:
+  // 1. Endpoint has been added as a favourite
+  // 2. Endpoint that has child favourites
+  // 3. Remaining endpoints
+  private orderEndpoints(endpoints: IRequestEntityTypeState<EndpointModel>, favorites: IUserFavoritesGroups): EndpointModel[] {
+    const processed = {};
+    const result = [];
+
+    Object.keys(favorites).forEach(fav => {
+      if (!favorites[fav].ethereal) {
+        const id = this.userFavoriteManager.getEndpointIDFromFavoriteID(fav);
+        if (!!endpoints[id] && !processed[id]) {
+          processed[id] = true;
+          result.push(endpoints[id]);
+        }
+      }
+    });
+
+    Object.keys(favorites).forEach(fav => {
+      if (favorites[fav].ethereal) {
+        const id = this.userFavoriteManager.getEndpointIDFromFavoriteID(fav);
+        if (!!endpoints[id] && !processed[id]) {
+          processed[id] = true;
+          result.push(endpoints[id]);
+        }
+      }
+    });
+
+    Object.values(endpoints).forEach(ep => {
+      if (!processed[ep.guid]) {
+        processed[ep.guid] = true;
+        result.push(ep);
+      }
+    })
+
+    // Filter out the disconnected ones
+    return result.filter(ep => ep.connectionStatus === 'connected');
+  }
+
+  // Automatic layout - select the best layout based on the available endpoints
+  private automaticLayout(): Observable<HomePageCardLayout> {
+    return this.endpointsService.endpoints$.pipe(
+      map(eps => Object.values(eps)),
+      map(eps => eps.filter(ep => {
+        const defn = entityCatalog.getEndpoint(ep.cnsi_type, ep.sub_type);
```

Comment:
Use one method to get the valid endpoints - this would include disconnected ones?